plenum/server/replica.py

```diff
@@ -806,28 +897,44 @@ def nonFinalisedReqs(self, reqKeys: List[Tuple[str, int]]):
         """
         return {key for key in reqKeys if not self.requests.isFinalised(key)}
 
-    def __is_next_pre_prepare(self, ppSeqNo: int):
-        if ppSeqNo != self.__last_pp_seq_no + 1:
+    def __is_next_pre_prepare(self, view_no: int, pp_seq_no: int):
+        if view_no == self.viewNo and pp_seq_no == 1:
+            # First PRE-PREPARE in a new view
+            return True
+
+        (last_pp_view_no, last_pp_seq_no) = self.__last_pp_3pc
+
+        if last_pp_view_no > view_no:
+            return False
+
+        if last_pp_view_no < view_no:
+            assert view_no == self.viewNo
+            last_pp_seq_no = 0
+
+        if pp_seq_no - last_pp_seq_no != 1:
             logger.debug('{} missing PRE-PREPAREs between {} and {}'.
-                         format(self, ppSeqNo, self.__last_pp_seq_no))
+                         format(self, pp_seq_no, last_pp_seq_no))
+            # TODO: think of a better way, urgently
```

Comment:
It does not affect the correcness, but it makes it look clunkly, lets have a low priority refactoring ticket